### PR TITLE
workflows: nix-ci: init; build&cache flake outputs

### PR DESCRIPTION
**Pull request by SomeoneSerge**
Things done
---

- [x] Adds jobs to build and publish binary cache for x86_64-linux, x86_64-darwin, and (hopefully, using github's accelerated KVM; not sure how the performance is going to work out) aarch64-linux
  - [ ] The job is only active when the "repository variable" `CACHIX_NAME` and the "secret" `CACHIX_AUTH_TOKEN` are configured.
- [x] Fixes the actions to publish the flake on flakestry.dev and flakehub.com
- [x] Adds a job that updates the `flake.lock` file once a week and opens a respective PR
- [ ] Suggests the binary cache via the `flake.nix`
  - [ ] Will finish once the repository variables and secrets are set up, i.e. once we've agreed on the address of the cache

Some example runs: https://github.com/SomeoneSerge/llama.cpp/actions/runs/7365553235/job/20046977979

@ggerganov I see you gave @philiptaron and me the commit access, thanks! I presume we may use it to merge Nix-specific changes faster?

I'd still like you to have a look at this PR because it touches the CI. I'd also need your help with setting up the variables, which mostly have to do with Cachix. That's a service which can host a public binary cache for Nix. You can authenticate at https://www.cachix.org/ using oauth and github

CC @philiptaron and @ParetoOptimalDev for review

How to use this
---

The `flake.nix` is going to export an extra `nixConfig` output, which would configure Nix to use additional "substituters" when this repo is used as a flake. In the interactive mode the Nix CLI tools ask users whether they want to trust and accept the extra config, and the "flake" can be used regardless of whether the users choose to accept the config.

The `flake.nix` is going to extend the `extra-trusted-public-keys`. It's the users' responsibility to compare these values to those on the respective cachix pages. A trusted substituter has the capacity to "fake" the output of any derivation in the nix store, which has implications to security. Consuming the binary cache, the user also implicitly trusts the github runners, and (currently) the additional cache with cuda variants of the dependencies.

Essentially, presume that the cache is there for development and tinkering purposes. In an environment where one could potentially access sensitive data, better reject the extra config and build from scratch or even set up your own cache.

### CODE REVIEW

File: .github/workflows/nix-ci.yml
Comment (SomeoneSerge) [inline .github/workflows/nix-ci.yml:11]: I could've added these jobs to the "CI" workflow, but I chose to make a separate one so we can have `*.nix` in the `paths`

File: .github/workflows/nix-ci.yml
Comment (SomeoneSerge) [inline .github/workflows/nix-ci.yml:29]: These two come from here:

image

...which are publicly visible on the corresponding cachix page:

image

File: .github/workflows/nix-ci.yml
Comment (SomeoneSerge) [inline .github/workflows/nix-ci.yml:65]: This one is private, set up in: 
image

...which you generate here:

image


image

File: .github/workflows/nix-ci.yml
Comment (SomeoneSerge) [inline .github/workflows/nix-ci.yml:32]: This uses github's internal cache to speed the workflows up (which is closer to the runners than any of the used cachix-es or cache.nixos.org). Since the internal cache is limited and there are many other CI jobs in this repo, we set `upstream-cache` to avoid caching things available in our own cachix

File: .github/workflows/nix-ci.yml
Comment (philiptaron) [inline .github/workflows/nix-ci.yml:8]: <details><summary>Currently, this excludes the following files.</summary>

```
LICENSE
README.md
SHA256SUMS
awq-py/README.md
awq-py/awq/apply_awq.py
awq-py/requirements.txt
build.zig
ci/README.md
ci/run.sh
cmake/FindSIMD.cmake
codecov.yml
common/build-info.cpp.in
convert-hf-to-gguf.py
convert-llama-ggml-to-gguf.py
convert-lora-to-ggml.py
convert-persimmon-to-gguf.py
convert.py
docs/BLIS.md
docs/llama-star/idea-arch.key
docs/llama-star/idea-arch.pdf
docs/token_generation_performance_tips.md
examples/Miku.sh
examples/alpaca.sh
examples/batched-bench/README.md
examples/batched/README.md
examples/chat-13B.sh
examples/chat-persistent.sh
examples/chat-vicuna.sh
examples/chat.sh
examples/convert-llama2c-to-ggml/README.md
examples/embedding/README.md
examples/export-lora/README.md
examples/finetune/README.md
examples/finetune/convert-finetune-checkpoint-to-gguf.py
examples/finetune/finetune.sh
examples/gpt4all.sh
examples/infill/README.md
examples/jeopardy/README.md
examples/jeopardy/graph.py
examples/jeopardy/jeopardy.sh
examples/jeopardy/qasheet.csv
examples/jeopardy/questions.txt
examples/json-schema-to-grammar.py
examples/llama-bench/README.md
examples/llama.swiftui/README.md
examples/llama.swiftui/llama.swiftui.xcodeproj/project.pbxproj
examples/llama.swiftui/llama.swiftui.xcodeproj/project.xcworkspace/contents.xcworkspacedata
examples/llama.swiftui/llama.swiftui/Assets.xcassets/AccentColor.colorset/Contents.json
examples/llama.swiftui/llama.swiftui/Assets.xcassets/AppIcon.appiconset/Contents.json
examples/llama.swiftui/llama.swiftui/Assets.xcassets/Contents.json
examples/llama.swiftui/llama.swiftui/Preview Content/Preview Assets.xcassets/Contents.json
examples/llama.vim
examples/llama2-13b.sh
examples/llama2.sh
examples/llava/README.md
examples/llava/convert-image-encoder-to-gguf.py
examples/llava/llava-surgery.py
examples/llm.vim
examples/lookahead/README.md
examples/lookup/README.md
examples/main-cmake-pkg/README.md
examples/main/README.md
examples/make-ggml.py
examples/parallel/README.md
examples/perplexity/README.md
examples/quantize/README.md
examples/reason-act.sh
examples/server-llama2-13B.sh
examples/server/README.md
examples/server/api_like_OAI.py
examples/server/chat-llama2.sh
examples/server/chat.mjs
examples/server/chat.sh
examples/server/deps.sh
examples/server/public/completion.js
examples/server/public/index.html
examples/server/public/index.js
examples/server/public/json-schema-to-grammar.mjs
examples/simple/README.md
examples/speculative/README.md
examples/train-text-from-scratch/README.md
examples/train-text-from-scratch/convert-train-checkpoint-to-gguf.py
flake.lock
gguf-py/LICENSE
gguf-py/README.md
gguf-py/examples/writer.py
gguf-py/gguf/__init__.py
gguf-py/gguf/constants.py
gguf-py/gguf/gguf.py
gguf-py/gguf/gguf_reader.py
gguf-py/gguf/gguf_writer.py
gguf-py/gguf/py.typed
gguf-py/gguf/tensor_mapping.py
gguf-py/gguf/vocab.py
gguf-py/pyproject.toml
gguf-py/scripts/__init__.py
gguf-py/scripts/gguf-convert-endian.py
gguf-py/scripts/gguf-dump.py
gguf-py/scripts/gguf-set-metadata.py
gguf-py/tests/test_gguf.py
grammars/README.md
grammars/arithmetic.gbnf
grammars/c.gbnf
grammars/chess.gbnf
grammars/japanese.gbnf
grammars/json.gbnf
grammars/json_arr.gbnf
grammars/list.gbnf
media/llama-leader.jpeg
media/llama0-banner.png
media/llama0-logo.png
media/llama1-banner.png
media/llama1-logo.png
mypy.ini
prompts/LLM-questions.txt
prompts/alpaca.txt
prompts/assistant.txt
prompts/chat-with-baichuan.txt
prompts/chat-with-bob.txt
prompts/chat-with-qwen.txt
prompts/chat-with-vicuna-v0.txt
prompts/chat-with-vicuna-v1.txt
prompts/chat.txt
prompts/dan-modified.txt
prompts/dan.txt
prompts/mnemonics.txt
prompts/parallel-questions.txt
prompts/reason-act.txt
requirements-hf-to-gguf.txt
requirements.txt
run_with_preset.py
scripts/LlamaConfig.cmake.in
scripts/build-info.cmake
scripts/build-info.sh
scripts/convert-gg.sh
scripts/gen-build-info-cpp.cmake
scripts/get-flags.mk
scripts/get-wikitext-2.sh
scripts/qnt-all.sh
scripts/run-all-perf.sh
scripts/run-all-ppl.sh
scripts/server-llm.sh
scripts/sync-ggml-am.sh
scripts/sync-ggml.last
scripts/sync-ggml.sh
scripts/verify-checksum-models.py
tests/test-tokenizer-0-falcon.py
tests/test-tokenizer-0-llama.py
```
</details>

As more Python gets packaged, I think we'll add Python files into this list of paths. Maybe the grammars too.

File: .github/workflows/nix-ci.yml
Comment (philiptaron) [reply .github/workflows/nix-ci.yml:32]: I think it's worth a bit. We can see its impact using the cache stats, I think.

File: .github/workflows/nix-ci.yml
Comment (philiptaron) [inline .github/workflows/nix-ci.yml:71]: Since the other variant assumes that this returns x86_64-linux, I think it's fine to write that out.

File: .github/workflows/nix-ci.yml
Comment (philiptaron) [inline .github/workflows/nix-ci.yml:69]: It looks like [`--skip-cached`](https://github.com/Mic92/nix-fast-build#avoiding-redundant-package-downloads) flag might not work that well with the Magic Nix Cache. But maybe it's fine.

File: .github/workflows/nix-flake-update.yml
Comment (philiptaron) [inline .github/workflows/nix-flake-update.yml:5]: Nice, this is a good schedule.

File: .github/workflows/nix-flakestry.yml
Comment (philiptaron) [inline .github/workflows/nix-flakestry.yml:29]: v4 elsewhere

File: flake.nix
Comment (philiptaron) [inline flake.nix:4]: @ggerganov ought to be in control of this, I think, unless he'd rather delegate.

File: flake.nix
Comment (philiptaron) [inline flake.nix:2]: I'd like this to be at the bottom of the file, but I think it definitely doesn't matter.

File: flake.nix
Comment (philiptaron) [inline flake.nix:122]: Ah, the aarch64-linux revealed this?

File: .github/workflows/nix-flakestry.yml
Comment (philiptaron) [inline .github/workflows/nix-flakestry.yml:1]: Maybe rename this to `nix-flake-publish.yml`?

**Review by philiptaron — APPROVED** [review]
I looked through all of this. Exciting. This puts the Nix-compiled artifacts in reach of basically everyone. Really nice.

File: flake.nix
Comment (SomeoneSerge) [reply flake.nix:4]: Yes, I meant to replace these once @ggerganov has set up another cachix

File: .github/workflows/nix-ci.yml
Comment (SomeoneSerge) [reply .github/workflows/nix-ci.yml:69]: H'm, why?

File: .github/workflows/nix-ci.yml
Comment (SomeoneSerge) [reply .github/workflows/nix-ci.yml:71]: This job runs over a matrix of `os \in [ "ubuntu-latest", "darwin-latest" ]`. Only the aarch64-linux variant is special-cased

File: flake.nix
Comment (SomeoneSerge) [reply flake.nix:122]: Yea just `nix flake show --all-systems` iirc

File: flake.nix
Comment (SomeoneSerge) [reply flake.nix:2]: Sure. Or I could make it alpabetical.

File: .github/workflows/nix-flake-update.yml
Comment (SomeoneSerge) [inline .github/workflows/nix-flake-update.yml:20]: @ggerganov just confirming, is it OK to add a label for nix-related issues/PRs?

File: .github/workflows/nix-ci.yml
Comment (philiptaron) [reply .github/workflows/nix-ci.yml:69]: Because I didn't understand it. After reading some code...

1. `--skip-cached` controls whether nix-eval-jobs receives the `--check-cache-status` flag
2. That flag makes nix-eval-jobs check if the derivations are present locally or in any configured substituters (i.e. binary cache). The information will be exposed in the `isCached` field of the JSON output.
3. If isCached is true, the derivation is not built.

That works just fine with the magic nix cache.

**Review by ggerganov — COMMENTED** [review]
> I presume we may use it to merge Nix-specific changes faster?

Yes, I'm not familiar at all with Nix / Flake so hoping for any kind of support on this as it seems something that is used. Feel free to push / merge Nix-related changes as you see fit. For other things - wait for an extra review.

Will wait for further instructions for setting Cachix - so far, I logged in.

**Comment by SomeoneSerge** [comment]
> ... Nix / Flake so hoping for any kind of support on this as it seems something that is used. Feel free to push / merge Nix-related changes as you see fit

Thanks, appreciated!

> Will wait for further instructions for setting Cachix - so far, I logged in.

1. "[Create a binary cache](https://app.cachix.org/cache)"
2. `->` "Pick a unique name" -> e.g. `llama-cpp` (would correspond to `llama-cpp.cachix.org`). This name goes into the `CACHIX_NAME` [repository variable](https://github.com/ggerganov/llama.cpp/settings/variables/actions) mentioned in https://github.com/ggerganov/llama.cpp/pull/4709#discussion_r1438708019.
3. There's going to be a blue ribbon with the "public key" at the top of the "[cache page](https://app.cachix.org/cache/llama-cpp/)". It goes into the `CACHIX_PUBLIC_KEY` [repository variable](https://github.com/ggerganov/llama.cpp/settings/variables/actions) mentioned in https://github.com/ggerganov/llama.cpp/pull/4709#discussion_r1438708019
4. `->` "[Push binaries](https://app.cachix.org/cache/llama-cpp#push)"
5. `->` "[Generate ...] [per cache auth tokens](https://app.cachix.org/cache/llama-cpp/settings/authtokens)"
6. `->` "Generate" `->` copy the private token into the [repositories' secrets](https://github.com/ggerganov/llama.cpp/settings/secrets/actions) as "`CACHIX_AUTH_TOKEN`", [as in the screenshot](https://github.com/ggerganov/llama.cpp/pull/4709#discussion_r1438708452)

That's the total of 2 "variables" and 1 "secret"

**Comment by ggerganov** [comment]
The 2 vars and 1 secret should now be configured: https://app.cachix.org/cache/llama-cpp

File: .github/workflows/nix-ci.yml
Comment (SomeoneSerge) [inline .github/workflows/nix-ci.yml:101]: This is using qemu/KVM and I'm not sure how slow this is going to be in practice. In my test repo I had one job that succeeded after 1h30m (oooof): https://github.com/SomeoneSerge/llama.cpp/actions/runs/7367025606/job/20049815302, and one job that had timed out after 5h: https://github.com/SomeoneSerge/llama.cpp/actions/runs/7365553235/job/20046977979

It's very much desirable to cache the outputs for jetsons (they're usually limited in storage and cuda requries lots of it at build time) but I have to wonder if it's reasonable to run on every PR/every push.

Maybe we could try and then revert? Or I could make this a separate on-schedule job. @ggerganov?

Comment (ggerganov) [reply .github/workflows/nix-ci.yml:101]: Ok, we can try and if it's too slow we can make a scheduled job

File: flake.nix
Comment (SomeoneSerge) [reply flake.nix:10]: Made it alphabetical, that OK?

File: .github/workflows/nix-publish-flake.yml
Comment (philiptaron) [inline .github/workflows/nix-publish-flake.yml:2]: & fh

**Review by philiptaron — APPROVED** [review]
Check it in!

**Review by philiptaron — APPROVED** [review]

**Review by philiptaron — APPROVED** [review]
Serge and I debugged the aarch64 failure to accidentally running the aarch64 version of cachix 😬

**Review by SomeoneSerge — COMMENTED** [review]
@philiptaron I think this can be merged now. I also think we should use the rebase&merge mode to keep the individual commits, in case we want to revert the jetson job quickly

The aarch64 regression has been addressed in the latest push, cf. a sample run: https://github.com/SomeoneSerge/llama.cpp/actions/runs/7371792412/job/20059516752

NB: The first run (or a few runs if overlapping in time) are going to be slower than the rest because of the cold cache

**Review by philiptaron — APPROVED** [review]
OK. The [proof of the pudding is in the tasting](https://github.com/SomeoneSerge/llama.cpp/actions/runs/7371792412), and I like the way that this pudding is tasting. Look at all that green!

**Comment by philiptaron** [comment]
I'm going to merge this once the [final step](https://github.com/SomeoneSerge/llama.cpp/actions/runs/7371792412/job/20059516752) completes.

**Pull request merged**
